src/optimproblem.cpp

```diff
@@ -897,6 +1066,63 @@ void OptimProblem::rollOut(Vec x){
       }
     }
   }
+
+}
+
+void OptimProblem::updateLagrangian(const double prev_mu, const Vec x, Vec lambda) {
+
+  evalF(x, lambda, true);
+
+  Vec lambda_incre;
+  VecCreate(PETSC_COMM_WORLD, &lambda_incre);
```

Comment:
To avoid allocating arrays at every call (slow) they could be allocated once in the constructor (faster)

Reply:
It might be. I found that the scalability is not achieved at > 30 windows, and this might have some contribution.

Reply:
I'll move allocations into the constructor when merging.